infor-design/enterprise-wc
# 1573- IdsDataGrid tree grid virtual scrolling refactor

**Explain the details for making this change. What existing problem does the pull request solve?**
This PR adds a virtual scroll/row recycling algorithm specific to virtual scroll on tree grids.
The current virtual scroll algo doesn't account for dynamic rows (hidden sub rows) in its calculation, which led to 
problems like a jumpy scrollbar, a scrollbar handle that doesn't track with the mouse, seemingly empty grid body with large nested datasets, and broken `scrollRowIntoView` and `rowStart` settings/methods. 

This PR adds tree virtual scrolling along side the default virtual scrolling, so the current working non-tree virtual scrolling should work as is.

**Related github/jira issue (required)**:
Fixes #1573 
Fixes #1587 

**Steps necessary to review your pull request (required)**:
1. Checkout branch
2. Go to http://localhost:4300/ids-data-grid/tree-grid-virtual-scroll-infinite-scroll.html

***Test dynamic row rendering***
1. Collapse/Expand any of the rows
2. Child rows that are collapsed/hidden should not be rendered in DOM (they're recycled to become other visible rows)
3. Child rows should keep their collapsed/expanded/selected states when their parents are expanded/collapsed
4. Number of rows in DOM should never exceed 100 (default max rows)
5. You can check dom row count via `$('ids-data-grid').rows.length`

***Check arias & row-index attributes***
1. Check that  grid row `aria-setsize`, `aria-level`, `ara-posinset`, `aria-expanded` are correct
2. The `row-index` and `data-index` of tree grid row attributes should reflect their actual data index in `$('ids-data-grid').data`

***Test infinite scrolling***
1. Scroll to the bottom to trigger an infinite scroll
3. The datagrid should not jump/shift up/down when infinite scroll is triggered
4. Test infinite scroll after clicking `Expand All`/`Collapse All`, there should be no `jumpiness`
5. Test that you are able to infinite scroll to the end of data. There are a total of `2100 rows` to load

***Test scrolling***
1. Check that the scrollbar size is correct on page load
2. Check that the scrollbar is correctly resized as you're expanding/collapsing and infinite scrolling rows
3. Scroll up/down the grid by dragging the scrollbar handle, the scroll handle should track with your mouse

***Test scrollRowIntoView***
1. For ease of testing, there's an added input at the top right of the [example page](http://localhost:4300/ids-data-grid/tree-grid-virtual-scroll-infinite-scroll.html) to `scrollRowIntoView()`
2. Input various row indexes and the grid should correctly scroll to that row
3. If the target row is within a collapsed parent, scrolling to that row should internally expand the parent row and make the target row visible

***Test `scrollMaxRows` setting***
1. This setting allows user to define virtual scroll max rows in dom. Default is 100
2. Change max rows via `$('ids-data-grid').scrollMaxRows = 50;`
3. The datagrid row count should immediately adjust to the new max scroll value

***Smoke test other virtual scroll examples***
Please give these a quick look to ensure they're still working ok
* http://localhost:4300/ids-data-grid/tree-grid.html
* http://localhost:4300/ids-data-grid/tree-grid-virtual-scroll.html
* http://localhost:4300/ids-data-grid/virtual-scroll.html
* http://localhost:4300/ids-data-grid/virtual-scroll-infinite-scroll.html
* http://localhost:4300/ids-data-grid/virtual-scroll-infinite-scroll-m3.html
* http://localhost:4300/ids-data-grid/virtual-scroll-row-start.html
* http://localhost:4300/ids-data-grid/tree-grid-expand-collapse.html

**Included in this Pull Request**:

**Additional**
There could be some optimizations made in the future to improve performance. 
For example, we could re introduce the use of `requestAnimationFrame` during scrolling.
For tree-grids, `get virtualRows()` could be cached and only updated when rows are expanded/collapsed/added/removed.
